tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -895,6 +898,203 @@ static void domain_remove_part_context_switch(void)
+	K_THREAD_STACK_MEMBER(stack, STEST_STACKSIZE);
+	int baz;
+};
+
+struct foo stest_member_stack;
+
+void z_impl_stack_info_get(u32_t *start_addr, u32_t *size)
+{
+	*start_addr = k_current_get()->stack_info.start;
+	*size = k_current_get()->stack_info.size;
+}
+
+Z_SYSCALL_HANDLER(stack_info_get, start_addr, size)
+{
+	Z_OOPS(Z_SYSCALL_MEMORY_WRITE(start_addr, sizeof(u32_t)));
+	Z_OOPS(Z_SYSCALL_MEMORY_WRITE(size, sizeof(u32_t)));
+
+	z_impl_stack_info_get((u32_t *)start_addr, (u32_t *)size);
+
+	return 0;
+}
+
+int z_impl_check_perms(void *addr, size_t size, int write)
+{
+	return z_arch_buffer_validate(addr, size, write);
+}
+
+Z_SYSCALL_HANDLER(check_perms, addr, size, write)
+{
+	return z_impl_check_perms((void *)addr, size, write);
+}
+
+void *z_impl_get_tls_pointer(void)
+{
+	return &(k_current_get()->userspace_local_data);
+}
+
+Z_SYSCALL_HANDLER(get_tls_pointer)
+{
+	return (u32_t)z_impl_get_tls_pointer();
+}
+
+void stack_buffer_scenarios(k_thread_stack_t *stack_obj, size_t obj_size)
+{
+	size_t stack_size;
+	u8_t val;
+	char *stack_start, *stack_ptr, *stack_end, *obj_start, *obj_end;
+	volatile char *pos;
+
+	expect_fault = false;
+
+
+	/* Dump interesting information */
+
+	stack_info_get((u32_t *)&stack_start, (u32_t *)&stack_size);
+	printk("   - Thread reports buffer %p size %zu\n", stack_start,
+	       stack_size);
+
+	stack_end = stack_start + stack_size;
+	obj_end = (char *)stack_obj + obj_size;
+	obj_start = (char *)stack_obj;
+
+	/* Assert that the created stack object, with the reserved data
+	 * removed, can hold a thread buffer of STEST_STACKSIZE
+	 */
+	zassert_true(STEST_STACKSIZE <= (obj_size - K_THREAD_STACK_RESERVED),
+		      "bad stack size in object");
+
+	/* Check that the stack info in the thread marks a region
+	 * completely contained within the stack object
+	 */
+	zassert_true(stack_end <= obj_end,
+		     "stack size in thread struct out of bounds (overflow)");
+	zassert_true(stack_start >= obj_start,
+		     "stack size in thread struct out of bounds (underflow)");
+
+	/* Check that the entire stack buffer is read/writable */
+	printk("   - check read/write to stack buffer\n");
+
+	/* Address of this stack variable is guaranteed to part of
+	 * the active stack, and close to the actual stack pointer.
+	 * Some CPUs have hardware stack overflow detection which
+	 * faults on memory access within the stack buffer but below
+	 * the stack pointer.
+	 *
+	 * First test does direct read & write starting at the estimated
+	 * stack pointer up to the highest addresses in the buffer
+	 */
+	stack_ptr = &val;
+	for (pos = stack_ptr; pos < stack_end; pos++) {
+		/* pos is volatile so this doesn't get optimized out */
+		val = *pos;
+		*pos = val;
+	}
+
+	if (z_arch_is_user_context()) {
+		/* If we're in user mode, check every byte in the stack buffer
+		 * to ensure that the thread has permissions on it.
+		 */
+		for (pos = stack_start; pos < stack_end; pos++) {
+			zassert_false(check_perms((void *)pos, 1, 0),
```

Comment:
It seems to me you are only checking read permissions here. EIther we need write permissions' checking too, or we should update the comment before, the the assert text....

Reply:
fixed